## Hardware Verification 2026-05-21 (Re-run)

**Hardware Context:**
- Machine: RTX 5070 Ti
- OS: Windows 11 Enterprise 10.0.26200
- Compiler: MSVC 17.14.10
- CUDA: 12.8
- Commit: `42d81b8` ("wire SMS probe harness for /verify gates 1 & 2")

**Build Result:** ❌ **BUILD FAILED** (cannot proceed to gates)

### Compilation Errors

#### 1. `pkg64_sms_probe.cu` lines 55-57:
```
error : expression preceding parentheses of apparent call must have (pointer-to-) function type
    float dx = gs.center.x - static_cast<float>(C_cpu.center.x());
                                                ^
```
**Root cause:** `Vec3::x` is a field, not a method. Should be `.x`, `.y`, `.z` (no parentheses).

#### 2. `cuda_renderer.cu` (line 62) and `pkg64_sms_probe.cu` (line 26):
```
error : namespace "astroray" has no member "Renderer"
    const astroray::Renderer& hostRenderer,
                    ^
```
**Root cause:** `Renderer` class is at global scope in `raytracer.h`, not in `astroray::` namespace. Should be `const Renderer&`.

### Phase 1 Gate Status

| Gate | Expected | Measured | Status |
|------|----------|----------|--------|
| Build | Clean | **FAILED** | ❌ BLOCKED |
| Gate 1: SMS CPU/GPU rel-err ≤ 1e-3 | PASS | - | ⏸️ NOT RUN |
| Gate 2: Caster flag upload | PASS | - | ⏸️ NOT RUN |
| Gate 3: GPU test suite (40 passed) | PASS | - | ⏸️ NOT RUN |

**Verdict:** Phase 1 probe harness wiring is incomplete. Commit `42d81b8` introduced syntax errors preventing compilation. Cannot proceed to hardware gate verification until build is clean.

**Delta vs previous run:** Previous run was blocked by missing probe harness pieces. This run has the probe file present but with compilation errors introduced during wiring.

--- a/module/blender_module.cpp
+++ b/module/blender_module.cpp
@@ -2600,9 +2600,10 @@ PYBIND11_MODULE(astroray, m) {
     // device probe harness to run against.
     m.def("build_bk7_sms_acceptance_scene", [](PyRenderer& r) {
         // Floor quad (two triangles) with lambertian grey.
+        py::dict floor_params;
         auto floor = r.createMaterial("lambertian",
             std::vector<float>{0.78f, 0.78f, 0.78f},
-            std::map<std::string, float>{});
+            floor_params);
         r.addTriangle(
             std::vector<float>{-2.4f, -1.2f, -2.2f},
             std::vector<float>{ 2.4f, -1.2f, -2.2f},
@@ -2614,14 +2615,18 @@ PYBIND11_MODULE(astroray, m) {
             std::vector<float>{-2.4f, -1.2f,  1.6f},
             floor);
         // Point light (sphere).
+        py::dict light_params;
+        light_params["intensity"] = 14.0f;
         auto light = r.createMaterial("light",
             std::vector<float>{1.0f, 1.0f, 1.0f},
-            std::map<std::string, float>{{"intensity", 14.0f}});
+            light_params);
         r.addSphere(std::vector<float>{0.0f, 1.6f, 1.0f}, 0.22f, light);
         // BK7 glass sphere (the caster).
+        py::dict glass_params;
+        glass_params["ior"] = 1.52f;
         auto glass = r.createMaterial("dielectric",
             std::vector<float>{1.0f, 1.0f, 1.0f},
-            std::map<std::string, float>{{"ior", 1.52f}});
+            glass_params);
         r.addSphere(std::vector<float>{0.0f, -0.4f, 0.15f}, 0.7f, glass);
         // Camera.
         r.setupCamera(

--- a/src/gpu/cuda_renderer.cu
+++ b/src/gpu/cuda_renderer.cu
@@ -59,7 +59,7 @@ float launchProfileLookup(int profileIndex, float lambda);
 
 // pkg64-gpu Phase 1 probe harness (defined in pkg64_sms_probe.cu).
 void launchPkg64SmsProbe(
-    const astroray::Renderer& hostRenderer,
+    const Renderer& hostRenderer,
     const GBVHNode*   d_bvhNodes,
     const GPrimitive* d_prims,
     const GTriangle*  d_tris,
@@ -136,7 +136,7 @@ struct CUDARenderer::Impl {
     int          profileCount = 0;
 
     // pkg64-gpu Phase 1 probe: stashed host Renderer for CPU reference.
-    const astroray::Renderer* hostRenderer = nullptr;
+    const Renderer* hostRenderer = nullptr;
 
     // Device info
     bool        available = false;

--- a/src/gpu/pkg64_sms_probe.cu
+++ b/src/gpu/pkg64_sms_probe.cu
@@ -23,7 +23,7 @@
 // whether the first flagged CPU caster has its isCausticCaster flag set on
 // the device side.
 void launchPkg64SmsProbe(
-    const astroray::Renderer& hostRenderer,
+    const Renderer& hostRenderer,
     const GBVHNode*   /*d_bvhNodes*/,
     const GPrimitive* /*d_prims*/,
     const GTriangle*  /*d_tris*/,
@@ -52,9 +52,9 @@ void launchPkg64SmsProbe(
             // Find the sphere matching C_cpu.center/radius.
             for (size_t i = 0; i < host_spheres.size(); ++i) {
                 const GSphere& gs = host_spheres[i];
-                float dx = gs.center.x - static_cast<float>(C_cpu.center.x());
-                float dy = gs.center.y - static_cast<float>(C_cpu.center.y());
-                float dz = gs.center.z - static_cast<float>(C_cpu.center.z());
+                float dx = gs.center.x - static_cast<float>(C_cpu.center.x);
+                float dy = gs.center.y - static_cast<float>(C_cpu.center.y);
+                float dz = gs.center.z - static_cast<float>(C_cpu.center.z);
                 float dr = gs.radius - C_cpu.radius;
                 if (dx*dx + dy*dy + dz*dz < 1e-4f && dr*dr < 1e-6f) {
                     // Found the caster sphere in d_spheres[i]. Check its flag.

--- a/tests/test_pkg64_gpu_sms_attempt_unit.py
+++ b/tests/test_pkg64_gpu_sms_attempt_unit.py
@@ -87,8 +87,17 @@ def _run_probe_subprocess() -> subprocess.CompletedProcess:
       [pkg64-gpu] sms attempt probe: ok=<0|1> fhero=<f> w=<f> tr=<f> \
           caster_flag_crossed=<0|1>
     """
+    from pathlib import Path
+    ROOT = Path(__file__).resolve().parent.parent
     script = textwrap.dedent(
         f"""
+        import sys
+        from pathlib import Path
+        sys.path.insert(0, {str(ROOT / 'tests')!r})
+        sys.path.insert(0, {str(ROOT)!r})
+        import runtime_setup
+        runtime_setup.configure_test_imports()
+
         import astroray
         # BK7-sphere acceptance scene — identical geometry/material to the
         # CPU SMS acceptance tests (tests/test_sms_caustic_validation.py).

